fields/types/gcsavatar/GcsAvatarType.js

```diff
@@ -0,0 +1,378 @@
+      return exists(this);
+    },
+    /**
+     * Resets the value of the field
+     *
+     * @api public
+     */
+    reset: function() {
+      reset(this);
+    },
+    /**
+     * Deletes the file from gcs and resets the field
+     *
+     * @api public
+     */
+    delete: function() {
+      var _this = this;
+      var promise = new Promise(function(resolve, reject) {
+        var gcsConfig = _this.gcsConfig;
+        var bucket = gcsHelper.initBucket(gcsConfig, _this.get(paths.gcsBucket));
+        var filename = _this.get(paths.filename);
+        if (filename && typeof filename === 'string') {
+          var filenameWithoutExt = filename.split('.')[0];
+          bucket.deleteFiles({
+            prefix: _this.get(paths.path) + filenameWithoutExt,
+          }, function(err) {
+            if (err) {
+              return reject(err);
+            }
+            resolve();
+          });
+        } else {
+          resolve();
+        }
+      });
+      reset(this);
+      return promise;
+    },
+  };
+
+  _.forEach(schemaMethods, function(fn, key) {
+    _this.underscoreMethod(key, fn);
+  });
+
+  // expose a method on the field to call schema methods
+  this.apply = function(item, method) {
+    return schemaMethods[method].apply(item, Array.prototype.slice.call(arguments, 2));
+  };
+
+  this.bindUnderscoreMethods();
+};
+
+
+/**
+ * Formats the field value
+ *
+ * @api public
+ */
+
+gcsavatar.prototype.format = function(item) {
+  if (this.hasFormatter()) {
+    return this.options.format(item, item[this.path]);
+  }
+  return item.get(this.paths.filename);
+};
+
+
+/**
+ * Detects the field have formatter function
+ *
+ * @api public
+ */
+
+gcsavatar.prototype.hasFormatter = function() {
+  return typeof this.options.format === 'function';
+};
+
+
+/**
+ * Detects whether the field has been modified
+ *
+ * @api public
+ */
+
+gcsavatar.prototype.isModified = function(item) {
+  return item.isModified(this.paths.filename);
+};
+
+
+/**
+ * Validates that a value for this field has been provided in a data object
+ *
+ * @api public
+ */
+
+gcsavatar.prototype.inputIsValid = function(data) { // eslint-disable-line no-unused-vars
+  // TODO - how should file field input be validated?
+  return true;
+};
+
+
+/**
+ * Updates the value for this field in the item from a data object
+ *
+ * @api public
+ */
+
+gcsavatar.prototype.updateItem = function(item, data, callback) { // eslint-disable-line no-unused-vars
+  // TODO - direct updating of data (not via upload)
+  process.nextTick(callback);
+};
+
+/**
+ * Uploads the file for this field
+ *
+ * @api public
+ */
+
+gcsavatar.prototype.uploadFile = function(item, file, update, callback) {
+  var _this = this;
+  var ONE_YEAR = 60 * 60 * 24 * 365;
+  var publicRead = _this.options.publicRead ? _this.options.publicRead : false;
+  var prefix = _this.options.datePrefix ? moment().format(_this.options.datePrefix) + '-' : '';
+  var filename = prefix + file.name;
```

Comment:
I suppose we don't need the prefix?

Reply:
Also, do we want to rename raw file here? (e.g hardcode the name to the email account prefix) The reason for doing this is that we don't need to deal with url reset in the cookie for the keystone-plugin.

Reply:
Yes, we could.

There's one little issue in this way, the file extension (e.g. .svg, .png) may not be the same with the previous one whether it is reset or not. We need to make keystone-plugin take the filename without extension.